pyomo/core/base/var.py

```diff
@@ -916,26 +919,38 @@ class VarList(IndexedVar):
     """
 
     def __init__(self, **kwds):
-        kwds['dense'] = False
+        #kwds['dense'] = False
         args = (Set(),)
-        self._nvars = 0
         IndexedVar.__init__(self, *args, **kwds)
 
     def construct(self, data=None):
         """Construct this component."""
         if __debug__ and logger.isEnabledFor(logging.DEBUG):
             logger.debug("Constructing variable list %s", self.name)
-        IndexedVar.construct(self, data)
+
+        # We need to ensure that the indices needed for initialization are
```

Comment:
I'm not quite sure what is going on in the logic below, but I'm worried that we're looping through all indices twice and constructing data up-front.  This sounds expensive.  I think there's a potential runtime and memory hit here.   Have you confirmed that that's not the case?

Can you comment on the difference between the previous design and the current one for Var?

Reply:
Yes.  The change is because we are now actually validating indices for VarList.  In the past, you could provide anything as keys when initializing a VarList using a dict and we would happily chug along.  Now that we are validating that VarList keys against the underlying index set, we need to make sure that the underlying index set is correct.  There is no impact on memory: you end up with the same number of variables as before.

Reply:
I see.  This is just for VarList, which makes sense.